src/main/kotlin/org/rust/ide/annotator/CargoCheckExtAnnotator.kt

```diff
@@ -0,0 +1,630 @@
+package org.rust.ide.annotator
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.TimeUnit
+import java.io.InputStream
+import java.lang.reflect.Type
+
+import kotlin.collections.*
+
+import com.intellij.lang.annotation.*
+import com.intellij.openapi.editor.*
+import com.intellij.psi.PsiFile
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.ProjectFileIndex
+import com.intellij.openapi.util.TextRange
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.openapi.vfs.newvfs.BulkFileListener
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent
+import com.intellij.openapi.diagnostic.Logger as IJLogger
+
+import com.google.gson.*
+
+import org.apache.commons.lang.StringEscapeUtils.escapeHtml
+
+object IJDebug {
+    val defaultTitle = "CARGO-CHECK"
+
+    fun log(msg : String, title : String = defaultTitle) {
+        IJLogger.getInstance(title).debug(msg)
+    }
+
+}
+
+object CargoJson {
```

Comment:
This object is **huge**. I think it is so because it tries to do two rather orthogonal things:

* parse json text into a typed Kotlin data structure,
* use that data structure to annotate the file.

Let's try to split into two parts. 

The first part, `CargoMessages`, should specify `data class`es and provide an ability to convert text output of cargo into Kotlin objects. It can live in a separate file and should be independent of the annotator (because it could be reused elsewhere). 

All the annotation logic should live in the annotator. 

So, the end result could look like this:

```Kotlin
for (line in cargoOutput) {
  val message = CargoMessage.parse(line)
  if (message.level == CargoMessageLevel.ERROR) {
    annotate(holder, message)
  }
}
```